Doctrine DBAL 3.2.0 blog post (#427)
doctrine/doctrine-website: source/blog/2021-11-26-dbal-3.2.0.md

Implementing this approach was impossible without introducing a new API which rendered the existing API obsolete.

Prior to DBAL `3.2.0`, the schema comparator could be only instantiated directly via the `new` keyword:

```php
$comparator = new Comparator();
```

Instantiated like this, the comparator doesn't have a notion of the target database platform and cannot perform the comparison properly. That is why, this way of instantiation is deprecated in favor of instantiating the comparator by the schema manager:

```php
$schemaManager = $connection->createSchemaManager();
$comparator = $schemaManager->createComparator();
```

This way, the schema manager can instantiate a platform-specific comparator and provide it with the necessary context (e.g. the default collation used by the database).

While the old API is still available, it is recommended to use the new API for more accurate comparison.

## Support for `psr/cache` ([#4620](https://github.com/doctrine/dbal/pull/4620))

> Since the Doctrine Cache library is being sunset, the new DBAL relase introduced the ability to use a PSR-6 compatible implementation for result caching.

Reviewer: ```suggestion
Since the Doctrine Cache library is being sunset, the new DBAL release introduced the ability to use a PSR-6 compatible implementation for result caching.
```

While both the `doctrine/cache` and `psr/cache` APIs will be supported until the next major DBAL release, we recommend users to switch to a PSR-6 compatible implementation in their projects.

## Support for `psr/log` ([#4967](https://github.com/doctrine/dbal/pull/4967))

The `SQLLogger` interface was designed long ago and has certain limitations: there is no way to log messages at different logger levels and it is really challenging to extend the logger functionality without introducing breaking API changes.

The new DBAL release introduces a new middleware that can delegate logging to a PSR-3 compatible implementation.

> Note, that the new logger won't produce the messages identical to the ones produced by the old one. If you have any processes built around analysing log messages, you may need to make some changes before adopting the new API.

Reviewer: ```suggestion
Note that the new logger won't produce the messages identical to the ones produced by the old one. If you have any processes built around analysing log messages, you may need to make some changes before adopting the new API.
```

## Always cache the full result ([#5003](https://github.com/doctrine/dbal/pull/5003))

The implementation of the result cache prior to DBAL 3.2.0 would store the result set in the cache only once it was fetched completely. It led to the following issues:

1. If the result isn't yet cached and its consumer didn't fetch it completely, the query would be executed again.

> 2. In case of a cache miss, the DBAL would call `get()` the cache twice: once to fetch the data and once to merge the just fetched result with other results that may be stored in the cache.

Reviewer: Note sure about that one:
```suggestion
2. In case of a cache miss, the DBAL would `get()` the cache entry twice: once to fetch the data and once to merge the just fetched result with other results that may be stored in the cache.
```

> The new implementation stores the results in the cache right after was fetched. It simplifies the caching layer significantly and makes its behavior more straightforward.

Reviewer: ```suggestion
The new implementation stores the results in the cache right after they were fetched. It simplifies the caching layer significantly and makes its behavior more straightforward.
```

## Add events for Transaction begin/commit/rollback ([#4622](https://github.com/doctrine/dbal/pull/4622))

The new DBAL version introduces three more transaction-related events:

- `onTransactionBegin`,
- `onTransactionCommit`,
- `onTransactionRollBack`.

> Subscribing to those might be helpful if, the application logic integrates the database transaction flow with the business logic implemented outside of the database. For instance, in the filesystem.

Reviewer: ```suggestion
Subscribing to those might be helpful if the application logic integrates the database transaction flow with the business logic implemented outside of the database. For instance, in the filesystem.
```

## Basic exception handling in IBM DB2 and SQL Server drivers ([#4929](https://github.com/doctrine/dbal/pull/4929), [#4928](https://github.com/doctrine/dbal/pull/4928))

> The DBAL provides a mechanism that converts driver-specific error codes to portable error-specific exceptions. For instance an attempt to insert NULL into a column that has a `NOT NULL` constraint applied will result in error with the code 1566 on MySQL and in ORA-01400 on Oracle. The DBAL will convert these two errors to a portable `NotNullConstraintViolationException`.

Reviewer: ```suggestion
The DBAL provides a mechanism that converts driver-specific error codes to portable error-specific exceptions. For instance an attempt to insert `NULL` into a column that has a `NOT NULL` constraint applied will result in error with the code `1566` on MySQL and in `ORA-01400` on Oracle. The DBAL will convert these two errors to a portable `NotNullConstraintViolationException`.
```

Historically, the DBAL drivers based on the `ibm_db2`, `sqlsrv` and `pdo_sqlsrv` extensions did not support this feature and would thow a generic `DriverException`.

As of DBAL 3.2.0, this feature is supported by all bundled drivers.

## Improved `AbstractPlatform::getLengthExpression()` ([#4855](https://github.com/doctrine/dbal/pull/4855))

Although the `LENGTH` expression was implemented for all supported database platforms, the different implementations didn't have consistent semantics:

> 1. Most of the implementation would return the length in characters (Unicode code points), which is the most expected behavior.

Reviewer: ```suggestion
1. Most implementations would return the length in characters (Unicode code points), which is the most expected behavior.
```
